Sources/Ollama/Client.swift

```diff
@@ -326,8 +333,102 @@ extension Client {
         if let template {
             params["template"] = .string(template)
         }
-
+        
+        if toolDefinitions.count > 0 {
+            params["tools"] = try Value(toolDefinitions)
+        }
+        
         return try await fetch(.post, "/api/chat", params: params)
+        
+    }
+    
+    /// Generates the next message in a chat with a provided model.
+    ///
+    /// - Parameters:
+    ///   - model: The name of the model to use for the chat.
+    ///   - messages: The messages of the chat, used to keep a chat memory.
+    ///   - options: Additional model parameters as specified in the Modelfile documentation.
+    ///   - tools: Override the list of tools to include in this chat.
+    ///   - template: The prompt template to use (overrides what is defined in the Modelfile).
+    /// - Returns: A `ChatResponse` containing the generated message and additional information.
+    /// - Throws: An error if the request fails or the response cannot be decoded.
+    public func autoRunChat(
```

Comment:
Assuming this functionality is accepted into the package, and the package becomes a "get up and running" easily package ([discussion](https://github.com/mattt/ollama-swift/pull/1#discussion_r1817843629)) _then (and only then)_ I personally would personally expect this functionality to be rolled into the base `chat` function.

It's confusing as a default to pass tools to the client / chat function and for nothing to happen with them.

If a user wanted to specialize the behaviour further we could offer different tool processing strategies e.g. interrupt, concurrent, serial.

Reply:
Base on @mattt's comments I've simplified the API, removing `self.tools`.

Based on @nicorichard's feedback about how he'd need to use the API and his PR's I've merged in the changes to tests to test the external API surface area, the fixes for the missing public access modifiers.

Also merged in a version of @nicorichard's suggestions moving the actions to be `async`.

I've also added a new `toolApproval` parameter. This async closure lets you approve or disallow each tool usage asynchronously (so can even hook up user input to accept each one, or just sensitive ones). The parameter is now autofilled by one that approves everything. Removing the approval closure (passing nil) makes it just return the raw chat itself, as does not passing in any tools, or the response not having any tool calls from the LLM.

The API is simpler now, more in keeping with the rest of the API surface area. And also lets you write async tools and approve/disallow each one now.

Reply:
Hoping this API is more in keeping with @mattt's API design style and fits better with the rest of the library.
While also giving you the missing bits of power you need to build things with @nicorichard.